lib/committee.js

```diff
@@ -0,0 +1,143 @@
+import createDebug from 'debug'
+
+/** @import {Measurement} from './preprocess.js' */
+/** @import {RetrievalResult, CommitteeCheckError} from './typings.js' */
+
+const debug = createDebug('spark:committee')
+
+const REQUIRED_COMMITTEE_SIZE = 30
+
+export class Committee {
+  /**
+   * @param {Pick<Measurement, 'cid' | 'minerId'>} retrievalTask
+   */
+  constructor ({ cid, minerId }) {
+    this.retrievalTask = { minerId, cid }
+
+    /** @type {Measurement[]} */
+    this.measurements = []
+
+    /** @type {string  | CommitteeCheckError} */
+    this.indexerResult = 'MAJORITY_NOT_FOUND'
+
+    /** @type {RetrievalResult} */
+    this.retrievalResult = 'MAJORITY_NOT_FOUND'
+  }
+
+  /**
+   * @param {number} requiredCommitteeSize
+   * @returns
+   */
+  evaluate (requiredCommitteeSize = REQUIRED_COMMITTEE_SIZE) {
+    debug(
+      'Evaluating task %o with %s accepted measurements',
+      this.retrievalTask,
+      this.measurements.length
+    )
+
+    if (!this.#checkCommitteeSize(requiredCommitteeSize)) return
+    if (!this.#checkMeasuredField('providerId')) return
+    if (!this.#checkMeasuredField('indexerResult')) return
+
+    // TODO
+    // - provider_address,
+    // - protocol,
+
+    if (!this.#checkMeasuredField('retrievalResult')) return
+
+    // TODO
+    // - status_code,
+    // - timeout,
+    // - car size
+    // - car checksum
+    // - car too large
```

Comment:
I agree the side effects are not great. However, your solution does not completely remove the need for side effects. If we find the majority, then we still need to modify the minority measurements.

Also, your proposal will make it difficult to interactively debug this function because you cannot step over individual checks in the condition. Unless the debugger can do that now?

Reply:
> I agree the side effects are not great. However, your solution does not completely remove the need for side effects. If we find the majority, then we still need to modify the minority measurements.

With my proposal above this multi step process might not be necessary any more

> Also, your proposal will make it difficult to interactively debug this function because you cannot step over individual checks in the condition. Unless the debugger can do that now?

The debugger should be able to do that yeah

Reply:
Marking this thread as resolved, the PR uses the compound-key approach now.